[1/5] Light client sync protocol

nicopernas opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the Altair spec logic defined in this version:

https://github.com/ethereum/consensus-specs/tree/208da34ac4e75337baf79adebf036ab595e39f15/specs/altair/light-client

It also implements the relevant endpoints specified at https://github.com/ethereum/beacon-APIs

https://github.com/prysmaticlabs/prysm/pull/11815

**Which issues(s) does this PR fix?**

Partially addresses https://github.com/prysmaticlabs/prysm/issues/11571

**Other notes for review**

This PR is a follow up to https://github.com/prysmaticlabs/prysm/pull/11815 which was too big to be reviewed. This is only the first PR in a series of 5 that I will be submitting sequentially. 

The heavy lifting here was done by [@qinlz2](https://github.com/qinlz2) and [@7AC](https://github.com/7AC), so kudos to them. 

-----
https://github.com/prysmaticlabs/prysm/pull/12854 is the last PR in the series that contains **all** changes so you can go back and see the full picture while reviewing this one.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=12853) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=12853) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prysmaticlabs/prysm?pullRequest=12853) it.</sub>

potuz reviewed (dismissed):
This looks good to me, requesting another approval cc @nisdas

terencechain commented on proto/eth/v2/beacon_lightclient.proto:
```diff
@@ -0,0 +1,96 @@
+import "proto/eth/v1/beacon_block.proto";
+import "proto/eth/v2/version.proto";
+import "proto/eth/v2/sync_committee.proto";
+//import "proto/engine/v1/execution_engine.proto";
```
Best to remove commented line if not used

```suggestion
```

terencechain commented on proto/eth/v2/beacon_lightclient.proto:
```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Prysmatic Labs.
```
```suggestion
// Copyright 2023 Prysmatic Labs.
```

rkapka commented:
Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: https://github.com/prysmaticlabs/prysm/issues/12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.

I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format. 

I think it would be most future-proof to split all types into 3 groups:
1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`

nicopernas commented:
@rkapka thanks for your suggestion. 

Are you suggesting we remove the grpc endpoints and have the http layer serving requests directly? if so, which would be a good example to follow? Is this the right doc to follow https://hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1?

> API-only structs defined in /beacon-chain/rpc/beacon/light_client_structs.go

These are HTTP API structs I would imagine right? If so, we added them to`beacon-chain/rpc/apimiddleware/structs.go` in a later PR. See https://github.com/prysmaticlabs/prysm/pull/12854. I guess it's only a matter of moving them to where you suggested. These should not support SSZ marshalling right?

rkapka commented:
> Are you suggesting we remove the grpc endpoints and have the http layer serving requests directly?

Yes

> if so, which would be a good example to follow?

https://github.com/prysmaticlabs/prysm/pull/12633/files is pretty simple. You basically have to implement the HTTP handler under `beacon-chain/rpc/eth/light-client` (I think that is a better separation then putting it under `/eth/beacon`) and register it in `beacon-chain/rpc/service.go`

> Is this the right doc to follow [hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1](https://hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1)?

This document is meant for a project where we rewrite **existing** grpc endpoints into HTTP, so it's not exactly what you want.

> These are HTTP API structs I would imagine right? If so, we added them tobeacon-chain/rpc/apimiddleware/structs.go in a later PR. See https://github.com/prysmaticlabs/prysm/pull/12854. I guess it's only a matter of moving them to where you suggested. These should not support SSZ marshalling right?

Right. `beacon-chain/rpc/eth/light-client/structs.go` will probably be the best place to put them. Using an HTTP handler means we can skip the middleware entirely, which makes things simpler and more performant. Also please note that requests/responses for HTTP handlers don't need the custom middleware tags, such as `hex="true"`.

nicopernas commented:
@rkapka we had also added support for the http streaming API (see [this commit](https://github.com/prysmaticlabs/prysm/pull/12854/commits/8835bec722c186029d0fa703ec39e4d2bf2f8a61) for example) 
Is that something you are also planning to remove? That code path uses proto messages and grpc, so I'm thinking that you are either reworking it o removing it entirely?

rkapka commented:
We will convert this to HTTP eventually, but extending it this way is fine for now.

nicopernas commented:
I have updated this and the "last PR of the series" (https://github.com/prysmaticlabs/prysm/pull/12854) following your suggestions. I've removed the GRPC endpoints and replaced them with pure HTTP handlers. I've also got rid of most of the middleware changes and reduced the number of messages defined in the protobuf. 

Hopefully this aligns better with your plans and makes for an easier transition out of grpc :) 

If I missed some any of your guidance (which is very likely) please let me know and I will happily update the patch.

rkapka commented on beacon-chain/rpc/eth/helpers/error_handling.go:
```diff
@@ -66,3 +68,38 @@ func PrepareStateFetchError(err error) error {
+	return nil
+}
+
+func HandleGetBlockErrorJson(blk interfaces.ReadOnlySignedBeaconBlock, err error) *http2.DefaultErrorJson {
```
Instead of creating this function, please use `WriteBlockFetchError` which is defined in `beacon-chain/rpc/eth/shared/errors.go`

nicopernas replied:
Yes! that function was not there when I wrote the initial refactor. Will use that one instead.

rkapka commented on beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+	// assert compute_epoch_at_slot(attested_state.slot) >= ALTAIR_FORK_EPOCH
+	attestedEpoch := types.Epoch(uint64(attestedState.Slot()) / uint64(config.SlotsPerEpoch))
+	if attestedEpoch < types.Epoch(config.AltairForkEpoch) {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid attested epoch: %d", attestedEpoch)
```
It would be better to return `RpcError` (defined in `beacon-chain/rpc/core/errors.go`) everywhere instead of `status` errors, which are a gRPC construct.

nicopernas replied:
I did remove the grpc status and codes from the handlers but these ones I missed. Will replace them.

nicopernas replied:
After replacing the grpc types with `core.RpcError` I got myself into a circular dependency:
- the LC helper depends on rpc.core
- rpc.core depends on blockchain
- blockchain depends on LC helper (this is added later in the series)

Easiest thing I can think of is moving the rpc/core/errors.go to a "submodule" like `rpc.core.errors`. That will only affect my lc helper and `beacon-chain/rpc/core/validator.go`. 

What do you think?

rkapka replied:
RPC is the outer layer of the system. It should not depend on the blockchain package. If your design requires this, then these helper functions should also be under blockchain. I think a new `blockchain/lightclient` package would make most sense. I haven't reviewed the logic of these functions yet, but they don't seem to be closely connected with any rpc functionality. Such a change would also mean that regular go errors will be returned from these functions.

nicopernas replied:
The LC helpers introduced in this PR do not depend on the blockchain package. The blockchain package will depend on the LC helpers in a future PR. 

However, `rpc.core` does depend on the `blockchain` package and that's where the circular dependency comes from. 

Not adding a dependency from the LC helpers on `rpc.core` will solve the issue. I think using regular errors will do the trick regardless of where the LC helpers live.

rkapka replied:
I want to avoid the blockchain package depending on anything in the rpc package, if possible.

rkapka commented on proto/eth/v2/custom.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+const (
+	NextSyncCommitteeIndex = uint64(55)
+	FinalizedRootIndex     = uint64(105)
```
do these consts have to be exported?

nicopernas replied:
They are used by the light client store later in the series.

rkapka commented on proto/eth/v2/custom.go:
```diff
@@ -0,0 +1,51 @@
+	return bytes.Equal(x.AggregatePubkey, other.AggregatePubkey)
+}
+
+func FloorLog2(x uint64) int {
```
does this function have to be exported?

nicopernas replied:
Also used by the light client store.

rkapka commented on proto/eth/v2/beacon_lightclient.proto:
```diff
@@ -0,0 +1,71 @@
+    uint64 signature_slot = 7 [(ethereum.eth.ext.cast_type) = "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives.Slot"];
+}
+
+message LightClientFinalityUpdateResponse {
```
This type can be defined under `rpc/eth/lightclient` because it doesn't need ssz

nicopernas replied:
This type is used in the http streaming api (see [here](https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/events/events.go#L217))  and `streamData()` expects a `proto.Message`.

rkapka replied:
I think we should differentiate between proto messages used in events and HTTP response structs that don't need to be proto messages. After all they are used in two different contexts. I would remove the `Response` suffix from protos and create the response structs next to HTTP handlers.

nicopernas replied:
I can remove the `Response` suffix, no prob. 

These types are only used in the events and not by the HTTP handlers. The objects that populate the HTTP handlers' responses are already defined in go structs next to the handlers as you initially suggested. See https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/light-client/structs.go

nicopernas replied:
I ended up replacing the `Response` suffix with `WithVersion`. The idea is that we wanted to follow the spec and have these methods: [`create_light_client_finality_update`](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_finality_update) and [`create_light_client_optimistic_update`](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_optimistic_update). The result is having the structs we have: `LightClientOptimisticUpdate` and `LightClientFinalityUpdate`. Adding a version to these "goes against" the spec. 

The other approach that we considered was deviating a bit from the spec and have more "efficient" types. For example, we could `LightClientUpdate` to represent both final and optimistic updates. Note that as of now, `LightClientUpdate` is a super set of `LightClientOptimisticUpdate` and `LightClientFinalityUpdate`. The single type could be returned from both finality and optimistic endpoints nullifying the correspondent fields in each case. While we are at it, we could throw the version in there too. 

Again, since we wanted to stay closer to the spec we went for the approach proposed in this PR. However, there's an argument to be made in favour of the second approach.

rkapka commented on proto/eth/v2/beacon_lightclient.proto:
```diff
@@ -0,0 +1,71 @@
+    uint64 signature_slot = 5 [(ethereum.eth.ext.cast_type) = "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives.Slot"];
+}
+
+message LightClientOptimisticUpdateResponse {
```
This type can be defined under `rpc/eth/lightclient` because it doesn't need ssz

notbdu commented:
> Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: #12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.
> 
> I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format, so it seems that protos are the best choice.
> 
> I think it would be most future-proof to split all types into 3 groups:
> 
> 1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
> 2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
> 3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`

What's the reason for the migration from gRPC to HTTP? Is it an issue w/ most downstream clients using HTTP and the perf overhead of the grpc gateway or something like that?

rkapka commented:
> > Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: #12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.
> > I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format, so it seems that protos are the best choice.
> > I think it would be most future-proof to split all types into 3 groups:
> > 
> > 1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
> > 2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
> > 3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`
> 
> What's the reason for the migration from gRPC to HTTP? Is it an issue w/ most downstream clients using HTTP and the perf overhead of the grpc gateway or something like that?

Hey! Yes, you are basically correct. Because of how Beacon APIs are specified, we require additional middleware to translate between HTTP and gRPC, which adds maintenance cost and significant overhead.

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	if syncAggregate.SyncCommitteeBits.Count() < config.MinSyncCommitteeParticipants {
+		return nil, fmt.Errorf("invalid sync committee bits count: %d", syncAggregate.SyncCommitteeBits.Count())
```
```suggestion
		return nil, fmt.Errorf("invalid sync committee bits count %d", syncAggregate.SyncCommitteeBits.Count())
```

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	if headerRoot != blockRoot {
+		return nil, fmt.Errorf("invalid header root: %v", headerRoot)
```
```suggestion
		return nil, fmt.Errorf("header root %#x not equal to block root %#x", headerRoot, blockRoot)
```

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+
+	// assert hash_tree_root(attested_header) == block.message.parent_root
+	attestedHeaderRoot, err := attestedHeader.HashTreeRoot()
+	if err != nil || attestedHeaderRoot != block.Block().ParentRoot() {
+		return nil, fmt.Errorf("invalid attested header root: %v", attestedHeaderRoot)
+	}
```
let's split this into two error messages - one for `err != nil` and one for non-equal roots

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+	attestedState state.BeaconState) (*ethpbv2.LightClientUpdate, error) {
+	// assert compute_epoch_at_slot(attested_state.slot) >= ALTAIR_FORK_EPOCH
+	attestedEpoch := types.Epoch(uint64(attestedState.Slot()) / uint64(config.SlotsPerEpoch))
+	if attestedEpoch < types.Epoch(config.AltairForkEpoch) {
```
You don't have to pass the config as a parameter. Use `params.BeaconConfig()` instead

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+			BodyRoot:      make([]byte, 32),
+		}
+
+		finalityBranch = make([][]byte, 6)
```
Please extract `6` to a constant with a meaningful name

rkapka commented on beacon-chain/blockchain/lightclient.go:
```diff
@@ -0,0 +1,240 @@
+			}
+
+			if finalizedHeaderRoot != bytesutil.ToBytes32(attestedState.FinalizedCheckpoint().Root) {
+				return nil, fmt.Errorf("invalid finalized header root: %v", finalizedHeaderRoot)
```
```suggestion
				return nil, fmt.Errorf("finalized header root %#x not equal to attested finalized checkpoint root %#x", finalizedHeaderRoot, bytesutil.ToBytes32(attestedState.FinalizedCheckpoint().Root)
```

rkapka reviewed (changes requested):
Both functions in `beacon-chain/blockchain/lightclient.go` don't have any tests. Please add a test for the happy case as a minimum.

rkapka commented:
One thing I don't see in the PR is the following part of the `create_light_client_update` function from the spec:
```
# `next_sync_committee` is only useful if the message is signed by the current sync committee
    if update_attested_period == update_signature_period:
        update.next_sync_committee = attested_state.next_sync_committee
        update.next_sync_committee_branch = compute_merkle_proof_for_state(
            attested_state, NEXT_SYNC_COMMITTEE_INDEX)
```
Is this on purpose?

nicopernas commented:
> One thing I don't see in the PR is the following part of the `create_light_client_update` function from the spec:
> 
> ```
> # `next_sync_committee` is only useful if the message is signed by the current sync committee
>     if update_attested_period == update_signature_period:
>         update.next_sync_committee = attested_state.next_sync_committee
>         update.next_sync_committee_branch = compute_merkle_proof_for_state(
>             attested_state, NEXT_SYNC_COMMITTEE_INDEX)
> ```
> 
> Is this on purpose?

The `create_light_client_update` implementation will be added by the next PR in the series. See https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/light-client/helpers.go#L95

I have resolved your latest comments just so I could track the ones I had left to fix. Will be working on the happy path testes now and send a new update.

rkapka commented:
Ah, ok. So these two functions are not implementations of any particular spec function?

nicopernas commented:
They are actually, it's just that we organised the code in a funny way. Not sure why though...

* `NewLightClientFinalityUpdateFromBeaconState` implements [create_light_client_finality_update](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_finality_update)
* `NewLightClientOptimisticUpdateFromBeaconState` implements [create_light_client_optimistic_update](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_optimistic_update)

See how `sendLightClientFinalityUpdate()` and `sendLightClientOptimisticUpdate()` use them in [process_block.go](https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/blockchain/process_block.go#L95) and you'll see what I mean. 

I will refactor and merge these to avoid this type of confusions.

rkapka commented:
I was actually asking about the `XXXFromBeaconState` functions. They look very similar to `create_light_client_update`, which confused me.

rkapka commented:
You will have to run `bazel run //:gazelle -- fix` to satisfy the linter

rkapka commented:
Looks like the branch has some conflicts. Happy to merge this in once it's resolved

nicopernas commented:
Thanks! I just rebased from latest `develop` and fixed the conflicts.

rkapka commented on beacon-chain/blockchain/lightclient_test.go:
```diff
@@ -0,0 +1,160 @@
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.ParentRoot, "Finalized header parent root is not zero")
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.StateRoot, "Finalized header state root is not zero")
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.BodyRoot, "Finalized header body root is not zero")
+	require.Equal(t, 6, len(update.FinalityBranch), "Invalid finality branch leaves")
```
```suggestion
	require.Equal(t, finalityBranchNumOfLeaves, len(update.FinalityBranch), "Invalid finality branch leaves")
```

rkapka reviewed (commented):
1 nitpick

nicopernas commented:
> 1 nitpick

Fixed and rebased from latest `develop`

rkapka commented:
> > 1 nitpick
> 
> Fixed and rebased from latest `develop`

Thanks a lot for this great work, I approved the PR. Looks like something else got merged in the meantime and the branch is out of date again :(

nicopernas commented:
> > > 1 nitpick
> > 
> > 
> > Fixed and rebased from latest `develop`
> 
> Thanks a lot for this great work, I approved the PR. Looks like something else got merged in the meantime and the branch is out of date again :(

How about now?
